core/src/main/java/com/datastax/oss/driver/internal/core/DefaultCluster.java

```diff
@@ -145,25 +149,63 @@ private void init() {
           .addContactPoints(initialContactPoints)
           .thenCompose(v -> context.topologyMonitor().init())
           .thenCompose(v -> metadataManager.refreshNodes())
-          .thenAccept(
-              v -> {
-                try {
-                  context.loadBalancingPolicyWrapper().init();
-                  context.configLoader().onDriverInit(context);
-                  LOG.debug("[{}] Initialization complete, ready", logPrefix);
-                  initFuture.complete(DefaultCluster.this);
-                  // TODO schedule full schema refresh asynchronously (does not block init)
-                } catch (Throwable throwable) {
-                  initFuture.completeExceptionally(throwable);
-                }
-              })
+          .thenAccept(this::afterInitialNodeListRefresh)
           .exceptionally(
               error -> {
                 initFuture.completeExceptionally(error);
                 return null;
               });
     }
 
+    private void afterInitialNodeListRefresh(@SuppressWarnings("unused") Void ignored) {
+      try {
+        boolean protocolWasForced =
+            context.config().getDefaultProfile().isDefined(CoreDriverOption.PROTOCOL_VERSION);
+        boolean needSchemaRefresh = true;
+        if (!protocolWasForced) {
+          ProtocolVersion currentVersion = context.protocolVersion();
+          ProtocolVersion bestVersion =
+              context
+                  .protocolVersionRegistry()
+                  .highestCommon(metadataManager.getMetadata().getNodes().values());
+          if (!currentVersion.equals(bestVersion)) {
+            LOG.info(
+                "[{}] Negotiated protocol version {} for the initial contact point, "
+                    + "but other nodes only support {}, downgrading",
+                logPrefix,
+                currentVersion,
+                bestVersion);
+            context.channelFactory().setProtocolVersion(bestVersion);
+            ControlConnection controlConnection = context.controlConnection();
+            // Might not have initialized yet if there is a custom TopologyMonitor
+            if (controlConnection.isInit()) {
+              controlConnection.reconnectNow();
+              // Reconnection already triggers a full schema refresh
+              needSchemaRefresh = false;
+            }
+          }
+        }
+        if (needSchemaRefresh) {
+          metadataManager.refreshSchema(SchemaElementKind.WHOLE_SCHEMA, null, null, null);
+        }
+        metadataManager.firstSchemaRefreshFuture().thenAccept(this::afterInitialSchemaRefresh);
```

Comment:
There is technically another very small possible timing issue (encountered more rarely):

In [`ControlConnection.onSuccessfulReconnect()`](https://github.com/datastax/java-driver/blob/java1295/core/src/main/java/com/datastax/oss/driver/internal/core/control/ControlConnection.java#L371), schemaRefresh is scheduled immediately, so it could happen before `refreshNodes` is completed. 

Wouldn't it be better to defer completion on cluster initialization until after `refreshNodes` completes and the LBP is initialized?